text/0000-cargo-asymmetric-tokens.md

```diff
@@ -0,0 +1,218 @@
+
+Maven Central [requires](https://maven.apache.org/repository/guide-central-repository-upload.html#pgp-signature) all uploads to be [signed by PGP](https://central.sonatype.org/publish/requirements/gpg/) and that the keys are registered with a public key server. Following the UNIX philosophy, they leave the actual act of signing up to independent implementations of PGP.
+It takes a lot of documentation to explain how to hook up all of these different parts to work together correctly. Furthermore, no assurance is made that the GPG signature and the Maven Central token used for upload represent the same identity.
+
+The npm client can pass along a `otp` option on the command line to act as [proof of 2FA](https://docs.npmjs.com/configuring-two-factor-authentication#sending-a-one-time-password-from-the-command-line). This provides a lot of the "over the wire" benefits of this RFC for the npm registry, but cannot be used by a third party after the fact to verify the uploaded identity nor is it easy for alternative registries to set up.
+
+[TUF](https://theupdateframework.io/) exclusively deals with how a client downloading packages through a mirror can be assured they came from a non-compromised copy of the registry. Which is not the problem this RFC is addressing.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How aggressively to push people off secret tokens? This RFC does not remove the existing use of secret tokens for publishing and yanking on private registries nor suggests a timeline for crates.io to use asymmetric tokens. There is an [RFC to allow authentication on more operations](https://github.com/rust-lang/rfcs/blob/f3aecb96eeb95542d81d6dc6b0a22c1245383604/text/0000-cargo-alternative-registry-auth.md), the expectation is that we will require the use of asymmetric tokens for this new functionality. This is a question that we will have to decide as we go through implementation and stabilization.
+
+What default settings should `cargo login --generate-keypair` use? What process should be used for changing these defaults as best practice changes? Where should it put the private keys?
+
+More generally, is all the user experience exactly correct for all the new flags? The expectation is that these will need to be changed and tweaked as we try using them after implementation.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+Figuring out how and when crates.io should support these kinds of tokens is left to a follow-up discussion/RFC. The motivation section describes some of the things that will need to be figured out.
+
+Only after crates.io is not using secret tokens should we remove the support for them in private registries (and the code in cargo to support it).
```

Comment:
I really like the idea of using public key crypto to improve cargo security, however I'm somewhat concerned about deprecating any other type of authentication. It may make it harder to integrate cargo into existing authentication systems.

Using a `credential-process` to issue a short-lived token seems like a flexible and secure option. It could work similar to how GitHub does HTTP authentication after installing the [credential provider](https://github.com/GitCredentialManager/git-credential-manager). Cargo would call the credential provider with the existing information (registry url, etc), the credential provider could verify identity any means it wanted (OAuth popup, 2FA key entry, etc), then produce a short-lived bearer token for cargo to send.

Reply:
I think you're reminding us that it is possible to use secret tokens with `credential-process` in a secure way. Which is definitely an important thing to keep in mind.

With some ingenuity, you can use a credential-process to smuggle a secret token into the format of a asymmetric token. The credential process creates a brand-new key pair, puts in "custom" the secret token and the public key. The server can validate that the asymmetric token is valid according to this RFC, but then makes all AuthN decisions based on the embedded secret token. The decision about removal is about how much work is involved and for whom, not what is possible.

Fundamentally, this is in the future possibilities section. This RFC does not suggest deprecation or removal. The time to discuss whether deprecation or removal is a good idea is in future RFC that suggests making that change.